### DISCUSSION TITLE
Installing WSL on a restricted Windows 10

### DISCUSSION THREAD

**Discussion by audetto** [microsoft/WSL#9984] [General]
At work, we live in a very restricted Windows 10 environment (19042.2728).

Basically no executables are allowed other than a limited set of folders.
Let's say `c:\dev` is one of them.

If I try to update wsl via `wsl --update --web-download` it will install it under my `c:\users\name` folder which is restricted.

Can I tell wsl to install elsewhere?

**Comment by audetto** [comment]
If I simply unzip the content of the released code from github, I get

"This process has no package identity"

**Comment by Biswa96** [comment]
> Let's say c:\dev is one of them.
> If I simply unzip the content of the released code from github

It is possible to install WSL from the GitHub release page in any directory. But it requires administrative permission. The WSL package register a WSLService which requires that higher level of permission. Without admin permission, the `Add-AppxPackage` command shows error like following.

```
add-appxpackage : Deployment failed with HRESULT: 0x80073CF6, Package could not be registered.
AppxManifest.xml(51,12): error 0x80073D28: Cannot register the MicrosoftCorporationII.WindowsSubsystemForLinux_1.2.3.0_x64__8wekyb3d8bbwe package. Administrator privileges required to install packaged service
```

**Comment by audetto** [comment]
Perfect

1. do you mind telling me how to specify the destination folder
2. and if things go wrong, how can I remove it (and continue to use the old non-uipdated version which is in `c:\windows\system32`) (I do not want to lock myself out of it completely and you can imagine the level of tech support I can get on this highly non-standard thing)

**Comment by audetto** [comment]
I don't think 1 is that easy:

See https://social.technet.microsoft.com/Forums/windowsserver/en-US/cdcffc5f-ad01-431c-a767-9acef8aaace9/question-on-addappxpackage-powershell-command?forum=winserverpowershell

or

https://stackoverflow.com/questions/52919268/edit-installation-path

Can I `register a WSLService` from the unzip folder manually?

**Comment by Biswa96** [comment]
> At work, we live in a very restricted Windows 10 environment (19042.2728).

You should ask someone to get the permission to install programs.

> do you mind telling me how to specify the destination folder

⚠️ Disclaimer: The following procedure is used for testing packages and should not be used for daily usage. The procedure cancels out the security measure provided by the Appx packaging.

1. Download the msixbundle package from release page (https://github.com/microsoft/WSL/releases).
2. Unzip it with any tool like 7zip, unzip etc. For example, Microsoft.WSL_1.2.3.0_x64_ARM64.msixbundle --> Microsoft.WSL_1.2.3.0_x64.msix.
3. Extract that msix file again with same tool in a directory. For example, Microsoft.WSL_1.2.3.0_x64.msix -> C:\WSL.
4. Register the package using Powershell as administrator. For example, `Add-AppxPackage -register C:\WSL\AppxManifest.xml`

**Reply by audetto** [reply]
>     4. Register the package using Powershell as administrator. For example, `Add-AppxPackage -register C:\WSL\AppxManifest.xml`

I am afraid it has not been possible

```
PS C:\Windows\System32> Add-AppxPackage -register C:\dev\wsl\AppxManifest.xml
Add-AppxPackage : Deployment failed with HRESULT: 0x80073CFF, To install this application you need either a Windows
developer license or a sideloading-enabled system.
Deployment of package MicrosoftCorporationII.WindowsSubsystemForLinux_1.1.3.0_x64__8wekyb3d8bbwe with package origin
Developer Signed failed because no valid license or sideloading policy could be applied. A developer license
(http://go.microsoft.com/fwlink/?LinkId=233074) or enterprise sideloading configuration
(http://go.microsoft.com/fwlink/?LinkId=231020) may be required.
NOTE: For additional information, look for [ActivityId] c320be0d-8673-0008-9c7d-27c37386d901 in the Event Log or use
the command line Get-AppPackageLog -ActivityID c320be0d-8673-0008-9c7d-27c37386d901
At line:1 char:1
+ Add-AppxPackage -register C:\dev\wsl\AppxManifest.xml
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (C:\dev\wsl\AppxManifest.xml:String) [Add-AppxPackage], Exception
    + FullyQualifiedErrorId : DeploymentError,Microsoft.Windows.Appx.PackageManager.Commands.AddAppxPackageCommand
```

Is there anything else I could try?

**Reply by Biswa96** [reply]
I am not sure if there is any restriction in your system. You could try to enable developer mode as explained here https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging

**Reply by audetto** [reply]
Unfortunately we have tons of restrictions & Developer Mode is already enabled.

**Reply by Biswa96** [reply]
Then you should ask the person or team who manages the system. Circumventing those rules may be illegal.

**Comment by audetto** [comment]
Thank you

> You should ask someone to get the permission to install programs.

There are 10 layers of contractors between me and the `someone`. Unfortunately, management does not understand the value of `wsl`.

**Comment by audetto** [comment]
One more question:

> `Add-AppxPackage -register C:\WSL\AppxManifest.xml`

What are the side effects of this?

- Does this interact with a previous installation of WSL?
- Can it be undone?
- Will it affect the existing non-packaged original wsl installation?
